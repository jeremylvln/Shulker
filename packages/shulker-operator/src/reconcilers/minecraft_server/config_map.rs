use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

use super::MinecraftServerReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ConfigMapBuilder {
    type OwnerType = MinecraftServer;
    type ResourceType = ConfigMap;

    fn name(minecraft_server: &Self::OwnerType) -> String {
        format!("{}-config", minecraft_server.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, minecraft_server: &Self::OwnerType) -> bool {
        minecraft_server
            .spec
            .config
            .existing_config_map_name
            .is_none()
    }

    async fn create(
        &self,
        minecraft_server: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerReconciler::get_common_labels(minecraft_server)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..ConfigMap::default()
        };

        Ok(config_map)
    }

    async fn update(
        &self,
        minecraft_server: &Self::OwnerType,
        config_map: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        config_map.data = Some(Self::get_data_from_spec(&minecraft_server.spec.config));
        Ok(())
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }

    pub fn get_data_from_spec(spec: &MinecraftServerConfigurationSpec) -> BTreeMap<String, String> {
        BTreeMap::from([
            (
                "init-fs.sh".to_string(),
                include_str!("../../../assets/server-init-fs.sh").to_string(),
            ),
            (
                "server.properties".to_string(),
                vanilla::VanillaProperties::from_spec(spec).to_string(),
            ),
            (
                "bukkit-config.yml".to_string(),
                bukkit::BukkitYml::from_spec(spec).to_string(),
            ),
            (
                "spigot-config.yml".to_string(),
                spigot::SpigotYml::from_spec(spec).to_string(),
            ),
            (
                "paper-global-config.yml".to_string(),
                paper::PaperGlobalYml::from_spec(spec).to_string(),
            ),
        ])
    }
}

#[cfg(test)]
mod tests {
    use crate::reconcilers::{
        builder::ResourceBuilder,
        minecraft_server::fixtures::{create_client_mock, TEST_SERVER},
    };

    #[test]
    fn name_contains_server_name() {
        // W
        let name = super::ConfigMapBuilder::name(&TEST_SERVER);

        // T
        assert_eq!(name, "my-server-config");
    }

    #[tokio::test]
    async fn create_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);

        // W
        let config_map = builder
            .create(&TEST_SERVER, "my-server-config")
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[tokio::test]
    async fn update_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let mut config_map = builder
            .create(&TEST_SERVER, "my-server-config")
            .await
            .unwrap();

        // W
        builder.update(&TEST_SERVER, &mut config_map).await.unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[test]
    fn get_data_from_spec_has_startup_script() {
        // G
        let spec = TEST_SERVER.spec.config.clone();

        // W
        let data = super::ConfigMapBuilder::get_data_from_spec(&spec);

        // T
        assert!(data.contains_key("init-fs.sh"));
    }

    #[test]
    fn get_data_from_spec_has_server_configs() {
        // G
        let spec = TEST_SERVER.spec.config.clone();

        // W
        let data = super::ConfigMapBuilder::get_data_from_spec(&spec);

        // T
        assert!(data.contains_key("server.properties"));
        assert!(data.contains_key("bukkit-config.yml"));
        assert!(data.contains_key("spigot-config.yml"));
        assert!(data.contains_key("paper-global-config.yml"));
    }
}

mod vanilla {
    use std::collections::BTreeMap;

    use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

    pub struct VanillaProperties(BTreeMap<String, String>);

    impl VanillaProperties {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            let mut properties = BTreeMap::new();

            properties.insert("max-players".to_string(), spec.max_players.to_string());
            properties.insert(
                "allow-nether".to_string(),
                (!spec.disable_nether).to_string(),
            );
            properties.insert("online-mode".to_string(), "false".to_string());
            properties.insert("prevent-proxy-connections".to_string(), "false".to_string());
            properties.insert("enforce-secure-profiles".to_string(), "true".to_string());

            VanillaProperties(properties)
        }
    }

    impl ToString for VanillaProperties {
        fn to_string(&self) -> String {
            let mut properties = String::new();

            for (key, value) in &self.0 {
                properties.push_str(&format!("{}={}\n", key, value));
            }

            properties
        }
    }

    #[cfg(test)]
    mod tests {
        use std::collections::BTreeMap;

        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::VanillaProperties::from_spec(&spec);

            // T
            insta::assert_debug_snapshot!(config.0);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::VanillaProperties(BTreeMap::from([(
                "my-str".to_string(),
                "my-value".to_string(),
            )]));

            // W
            let properties = config.to_string();

            // T
            insta::assert_snapshot!(properties);
        }
    }
}

mod bukkit {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::MinecraftServerConfigurationSpec;

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitYml {
        settings: BukkitSettingsYml,
        auto_updater: BukkitAutoUpdaterYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitSettingsYml {
        allow_end: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct BukkitAutoUpdaterYml {
        enabled: bool,
    }

    impl BukkitYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            BukkitYml {
                settings: BukkitSettingsYml {
                    allow_end: !spec.disable_end,
                },
                auto_updater: BukkitAutoUpdaterYml { enabled: false },
            }
        }
    }

    impl ToString for BukkitYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::BukkitYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::BukkitYml {
                settings: super::BukkitSettingsYml { allow_end: true },
                auto_updater: super::BukkitAutoUpdaterYml { enabled: false },
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}

mod spigot {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::{
        MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotYml {
        settings: SpigotSettingsYml,
        advancements: SpigotSaveableYml,
        players: SpigotSaveableYml,
        stats: SpigotSaveableYml,
        save_user_cache_on_stop_only: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotSettingsYml {
        bungeecord: bool,
        restart_on_crash: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct SpigotSaveableYml {
        disable_saving: bool,
    }

    impl SpigotYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            SpigotYml {
                settings: SpigotSettingsYml {
                    bungeecord: spec.proxy_forwarding_mode
                        == MinecraftServerConfigurationProxyForwardingMode::BungeeCord,
                    restart_on_crash: false,
                },
                advancements: SpigotSaveableYml {
                    disable_saving: true,
                },
                players: SpigotSaveableYml {
                    disable_saving: true,
                },
                stats: SpigotSaveableYml {
                    disable_saving: true,
                },
                save_user_cache_on_stop_only: true,
            }
        }
    }

    impl ToString for SpigotYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::SpigotYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::SpigotYml {
                settings: super::SpigotSettingsYml {
                    bungeecord: false,
                    restart_on_crash: false,
                },
                advancements: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                players: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                stats: super::SpigotSaveableYml {
                    disable_saving: true,
                },
                save_user_cache_on_stop_only: true,
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}

mod paper {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::minecraft_server::{
        MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalYml {
        proxies: PaperGlobalProxiesYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesYml {
        bungee_cord: PaperGlobalProxiesBungeeCordYml,
        velocity: PaperGlobalProxiesVelocityYml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesBungeeCordYml {
        online_mode: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct PaperGlobalProxiesVelocityYml {
        enabled: bool,
        online_mode: bool,
        secret: String,
    }

    impl PaperGlobalYml {
        pub fn from_spec(spec: &MinecraftServerConfigurationSpec) -> Self {
            PaperGlobalYml {
                proxies: PaperGlobalProxiesYml {
                    bungee_cord: PaperGlobalProxiesBungeeCordYml {
                        online_mode: spec.proxy_forwarding_mode
                            == MinecraftServerConfigurationProxyForwardingMode::BungeeCord,
                    },
                    velocity: PaperGlobalProxiesVelocityYml {
                        enabled: spec.proxy_forwarding_mode
                            == MinecraftServerConfigurationProxyForwardingMode::Velocity,
                        online_mode: true,
                        secret: "${CFG_VELOCITY_FORWARDING_SECRET}".to_string(),
                    },
                },
            }
        }
    }

    impl ToString for PaperGlobalYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use crate::reconcilers::minecraft_server::fixtures::TEST_SERVER;

        #[test]
        fn from_spec() {
            // G
            let spec = TEST_SERVER.spec.config.clone();

            // W
            let config = super::PaperGlobalYml::from_spec(&spec);

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::PaperGlobalYml {
                proxies: super::PaperGlobalProxiesYml {
                    bungee_cord: super::PaperGlobalProxiesBungeeCordYml { online_mode: false },
                    velocity: super::PaperGlobalProxiesVelocityYml {
                        enabled: true,
                        online_mode: true,
                        secret: "my-secret".to_string(),
                    },
                },
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}
