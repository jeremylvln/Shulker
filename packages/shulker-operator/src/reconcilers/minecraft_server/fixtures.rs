use std::collections::BTreeMap;

use http::{Request, Response};
use k8s_openapi::api::core::v1::EnvVar;
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServer, MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_SERVER: MinecraftServer = MinecraftServer {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            tags: vec!["lobby".to_string()],
            version: MinecraftServerVersionSpec {
                channel: MinecraftServerVersion::Paper,
                name: "1.20.1".to_string(),
                ..MinecraftServerVersionSpec::default()
            },
            config: MinecraftServerConfigurationSpec {
                existing_config_map_name: None,
                world: Some(ResourceRefSpec {
                    url: Some("https://example.com/my_world.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }),
                plugins: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_plugin.jar".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                patches: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_patch.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                max_players: 42,
                disable_nether: false,
                disable_end: true,
                server_properties: None,
                proxy_forwarding_mode: MinecraftServerConfigurationProxyForwardingMode::Velocity,
                lifecycle_strategy:
                    MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                skip_agent_download: false
            },
            pod_overrides: Some(MinecraftServerPodOverridesSpec {
                image: None,
                env: Some(vec![EnvVar {
                    name: "EXTRA_ENV".to_string(),
                    value: Some("my_value".to_string()),
                    ..EnvVar::default()
                }]),
                resources: None,
                affinity: None,
                node_selector: Some(BTreeMap::from([(
                    "beta.kubernetes.io/os".to_string(),
                    "linux".to_string()
                )])),
                tolerations: None,
                service_account_name: None,
                volume_mounts: None,
                volumes: None,
                ports: None,
                health: None,
            })
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
