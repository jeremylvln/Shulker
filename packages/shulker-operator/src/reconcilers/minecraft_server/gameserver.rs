use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::Capabilities;
use k8s_openapi::api::core::v1::ConfigMapVolumeSource;
use k8s_openapi::api::core::v1::Container;
use k8s_openapi::api::core::v1::EmptyDirVolumeSource;
use k8s_openapi::api::core::v1::EnvVar;
use k8s_openapi::api::core::v1::EnvVarSource;
use k8s_openapi::api::core::v1::ObjectFieldSelector;
use k8s_openapi::api::core::v1::PodSpec;
use k8s_openapi::api::core::v1::PodTemplateSpec;
use k8s_openapi::api::core::v1::SecretKeySelector;
use k8s_openapi::api::core::v1::SecurityContext;
use k8s_openapi::api::core::v1::Volume;
use k8s_openapi::api::core::v1::VolumeMount;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use crate::resources::resourceref_resolver::ResourceRefResolver;
use google_agones_crds::v1::game_server::GameServer;
use google_agones_crds::v1::game_server::GameServerEvictionSpec;
use google_agones_crds::v1::game_server::GameServerHealthSpec;
use google_agones_crds::v1::game_server::GameServerPortSpec;
use google_agones_crds::v1::game_server::GameServerSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerSpec;

use super::config_map::ConfigMapBuilder;
use super::MinecraftServerReconciler;

const MINECRAFT_SERVER_SHULKER_CONFIG_DIR: &str = "/mnt/shulker/config";
const MINECRAFT_SERVER_CONFIG_DIR: &str = "/config";
const MINECRAFT_SERVER_DATA_DIR: &str = "/data";
const MINECRAFT_SERVER_SHULKER_PROXY_AGENT_VERSION: &str = env!("CARGO_PKG_VERSION");
const MINECRAFT_SERVER_SHULKER_MAVEN_REPOSITORY: &str =
    "https://maven.jeremylvln.fr/artifactory/shulker";

lazy_static! {
    static ref PROXY_SECURITY_CONTEXT: SecurityContext = SecurityContext {
        allow_privilege_escalation: Some(false),
        read_only_root_filesystem: Some(true),
        run_as_non_root: Some(true),
        run_as_user: Some(1000),
        capabilities: Some(Capabilities {
            drop: Some(vec!["ALL".to_string()]),
            ..Capabilities::default()
        }),
        ..SecurityContext::default()
    };
}

pub struct GameServerBuilder {
    client: Client,
    resourceref_resolver: ResourceRefResolver,
}

#[async_trait::async_trait]
impl ResourceBuilder for GameServerBuilder {
    type OwnerType = MinecraftServer;
    type ResourceType = GameServer;

    fn name(minecraft_server: &Self::OwnerType) -> String {
        minecraft_server.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let game_server = GameServer {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerReconciler::get_common_labels(minecraft_server)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: GameServerSpec::default(),
            status: None,
        };

        Ok(game_server)
    }

    async fn update(
        &self,
        minecraft_server: &Self::OwnerType,
        game_server: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        game_server.spec =
            Self::get_game_server_spec(&self.resourceref_resolver, minecraft_server).await?;
        Ok(())
    }
}

impl GameServerBuilder {
    pub fn new(client: Client) -> Self {
        GameServerBuilder {
            client: client.clone(),
            resourceref_resolver: ResourceRefResolver::new(client.clone()),
        }
    }

    pub async fn get_game_server_spec(
        resourceref_resolver: &ResourceRefResolver,
        minecraft_server: &MinecraftServer,
    ) -> Result<GameServerSpec, anyhow::Error> {
        let pod_template_spec =
            Self::get_pod_template_spec(resourceref_resolver, minecraft_server).await?;

        let game_server_spec = GameServerSpec {
            ports: Some(vec![GameServerPortSpec {
                name: "minecraft".to_string(),
                container_port: 25565,
                protocol: "TCP".to_string(),
            }]),
            eviction: Some(GameServerEvictionSpec {
                safe: "OnUpgrade".to_string(),
            }),
            health: Some(GameServerHealthSpec {
                disabled: Some(false),
                initial_delay_seconds: Some(30),
                period_seconds: Some(15),
                failure_threshold: Some(5),
            }),
            template: pod_template_spec,
        };

        Ok(game_server_spec)
    }

    async fn get_pod_template_spec(
        resourceref_resolver: &ResourceRefResolver,
        minecraft_server: &MinecraftServer,
    ) -> Result<PodTemplateSpec, anyhow::Error> {
        let mut pod_spec = PodSpec {
            init_containers: Some(vec![Container {
                image: Some("alpine:latest".to_string()),
                name: "init-fs".to_string(),
                command: Some(vec![
                    "sh".to_string(),
                    format!("{}/init-fs.sh", MINECRAFT_SERVER_SHULKER_CONFIG_DIR),
                ]),
                env: Some(Self::get_init_env(resourceref_resolver, &minecraft_server.spec).await?),
                security_context: Some(PROXY_SECURITY_CONTEXT.clone()),
                volume_mounts: Some(vec![
                    VolumeMount {
                        name: "shulker-config".to_string(),
                        mount_path: MINECRAFT_SERVER_SHULKER_CONFIG_DIR.to_string(),
                        read_only: Some(true),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-config".to_string(),
                        mount_path: MINECRAFT_SERVER_CONFIG_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                ]),
                ..Container::default()
            }]),
            containers: vec![Container {
                image: Some("itzg/minecraft-server:2022.16.0-java17".to_string()),
                name: "minecraft-server".to_string(),
                env: Some(Self::get_env(&minecraft_server.spec)),
                image_pull_policy: Some("IfNotPresent".to_string()),
                security_context: Some(PROXY_SECURITY_CONTEXT.clone()),
                volume_mounts: Some(vec![
                    VolumeMount {
                        name: "server-config".to_string(),
                        mount_path: MINECRAFT_SERVER_CONFIG_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-data".to_string(),
                        mount_path: MINECRAFT_SERVER_DATA_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-tmp".to_string(),
                        mount_path: "/tmp".to_string(),
                        ..VolumeMount::default()
                    },
                ]),
                ..Container::default()
            }],
            service_account_name: Some(format!(
                "{}-server",
                &minecraft_server.spec.cluster_ref.name
            )),
            restart_policy: Some("Never".to_string()),
            volumes: Some(vec![
                Volume {
                    name: "shulker-config".to_string(),
                    config_map: Some(ConfigMapVolumeSource {
                        name: Some(ConfigMapBuilder::name(minecraft_server)),
                        ..ConfigMapVolumeSource::default()
                    }),
                    ..Volume::default()
                },
                Volume {
                    name: "server-config".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
                Volume {
                    name: "server-data".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
                Volume {
                    name: "server-tmp".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
            ]),
            ..PodSpec::default()
        };

        if let Some(pod_overrides) = &minecraft_server.spec.pod_overrides {
            if let Some(image_overrides) = &pod_overrides.image {
                pod_spec.containers[0].image = Some(image_overrides.name.clone());
                pod_spec.containers[0].image_pull_policy =
                    Some(image_overrides.pull_policy.clone());

                if let Some(image_pull_secrets) = pod_spec.image_pull_secrets.as_mut() {
                    image_pull_secrets.append(&mut image_overrides.image_pull_secrets.clone());
                }
            }

            if let Some(resources_overrides) = &pod_overrides.resources {
                pod_spec.containers[0].resources = Some(resources_overrides.clone());
            }

            if let Some(affinity_overrides) = &pod_overrides.affinity {
                pod_spec.affinity = Some(affinity_overrides.clone());
            }

            if let Some(node_selector_overrides) = &pod_overrides.node_selector {
                pod_spec.node_selector =
                    Some(node_selector_overrides.clone().into_iter().collect());
            }

            pod_spec.tolerations = pod_overrides.tolerations.clone();
        }

        Ok(PodTemplateSpec {
            metadata: Some(ObjectMeta {
                labels: Some(MinecraftServerReconciler::get_common_labels(
                    minecraft_server,
                )),
                annotations: Some(BTreeMap::<String, String>::from([(
                    "minecraftserver.shulkermc.io/tags".to_string(),
                    minecraft_server.spec.tags.join(","),
                )])),
                ..ObjectMeta::default()
            }),
            spec: Some(pod_spec),
        })
    }

    async fn get_init_env(
        resourceref_resolver: &ResourceRefResolver,
        spec: &MinecraftServerSpec,
    ) -> Result<Vec<EnvVar>, anyhow::Error> {
        let mut env: Vec<EnvVar> = vec![
            EnvVar {
                name: "SHULKER_CONFIG_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_CONFIG_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SERVER_CONFIG_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_CONFIG_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SERVER_DATA_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_DATA_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "TYPE".to_string(),
                value: Some(Self::get_type_from_version_channel(&spec.version.channel)),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_SERVER_AGENT_VERSION".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_PROXY_AGENT_VERSION.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_MAVEN_REPOSITORY".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_MAVEN_REPOSITORY.to_string()),
                ..EnvVar::default()
            },
        ];

        if let Some(world) = &spec.config.world {
            let url = resourceref_resolver.resolve(world).await?.as_url()?;

            env.push(EnvVar {
                name: "SERVER_WORLD_URL".to_string(),
                value: Some(url.to_string()),
                ..EnvVar::default()
            })
        }

        if let Some(plugins) = &spec.config.plugins {
            let urls: Vec<String> = resourceref_resolver
                .resolve_all(plugins)
                .await?
                .into_iter()
                .map(|url| url.to_string())
                .collect();

            env.push(EnvVar {
                name: "SERVER_PLUGIN_URLS".to_string(),
                value: Some(urls.join(",")),
                ..EnvVar::default()
            })
        }

        if let Some(patches) = &spec.config.patches {
            let urls: Vec<String> = resourceref_resolver
                .resolve_all(patches)
                .await?
                .into_iter()
                .map(|url| url.to_string())
                .collect();

            env.push(EnvVar {
                name: "SERVER_PATCH_URLS".to_string(),
                value: Some(urls.join(",")),
                ..EnvVar::default()
            })
        }

        Ok(env)
    }

    fn get_env(spec: &MinecraftServerSpec) -> Vec<EnvVar> {
        let mut env: Vec<EnvVar> = vec![
            EnvVar {
                name: "SHULKER_SERVER_NAME".to_string(),
                value_from: Some(EnvVarSource {
                    field_ref: Some(ObjectFieldSelector {
                        field_path: "metadata.name".to_string(),
                        ..ObjectFieldSelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_SERVER_NAMESPACE".to_string(),
                value_from: Some(EnvVarSource {
                    field_ref: Some(ObjectFieldSelector {
                        field_path: "metadata.namespace".to_string(),
                        ..ObjectFieldSelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "TYPE".to_string(),
                value: Some(Self::get_type_from_version_channel(&spec.version.channel)),
                ..EnvVar::default()
            },
            EnvVar {
                name: Self::get_version_env_from_version_channel(&spec.version.channel),
                value: Some(spec.version.name.clone()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "EULA".to_string(),
                value: Some("TRUE".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "COPY_CONFIG_DEST".to_string(),
                value: Some(MINECRAFT_SERVER_DATA_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SYNC_SKIP_NEWER_IN_DESTINATION".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SKIP_SERVER_PROPERTIES".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "REPLACE_ENV_IN_PLACE".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "REPLACE_ENV_VARIABLE_PREFIX".to_string(),
                value: Some("CFG_".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "CFG_VELOCITY_FORWARDING_SECRET".to_string(),
                value_from: Some(EnvVarSource {
                    secret_key_ref: Some(SecretKeySelector {
                        name: Some(format!("{}-forwarding-secret", spec.cluster_ref.name)),
                        ..SecretKeySelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "MEMORY".to_string(),
                value: Some("".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "JVM_XX_OPTS".to_string(),
                value: Some("-XX:MaxRAMPercentage=75".to_string()),
                ..EnvVar::default()
            },
        ];

        if let Some(pod_overrides) = &spec.pod_overrides {
            if let Some(env_overrides) = &pod_overrides.env {
                env.extend(env_overrides.clone());
            }
        }

        env
    }

    fn get_type_from_version_channel(channel: &str) -> String {
        match channel {
            "Bukkit" => "BUKKIT".to_string(),
            "Spigot" => "SPIGOT".to_string(),
            "Paper" => "PAPER".to_string(),
            _ => unreachable!("Unknown version channel"),
        }
    }

    fn get_version_env_from_version_channel(_channel: &str) -> String {
        "VERSION".to_string()
    }
}
