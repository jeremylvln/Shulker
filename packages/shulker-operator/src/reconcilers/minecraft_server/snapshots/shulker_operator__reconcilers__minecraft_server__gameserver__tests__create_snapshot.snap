---
source: packages/shulker-operator/src/reconcilers/minecraft_server/gameserver.rs
expression: game_server
---
apiVersion: agones.dev/v1
kind: GameServer
metadata:
  labels:
    app.kubernetes.io/component: minecraftserver
    app.kubernetes.io/name: my-server
    minecraftcluster.shulkermc.io/name: my-cluster
  name: my-server
  namespace: default
spec:
  ports: ~
  health: ~
  template: {}
  eviction: ~

